### PR TITLE
feat: implement `--forbidden-keys`

### PR DESCRIPTION
**Pull request by tigrato**
This PR implements a linter rule to deny usages of certain keys as attributes to enforce that developers don't override reserved keys managed by custom handlers.

### CODE REVIEW

**Comment by tmzane** [comment]
Hello, thank you for the PR, I like the idea!

I noticed that you first named it `--forbidden-key`, which I actually like better, because there might be other reasons why a log key is forbidden (e.g., a styleguide). In general, I think `sloglint` does not need to know the semantics behind the keys, it just needs to forbid them.

**Comment by tigrato** [comment]
@tmzane thanks.

I reverted the feature to its original name `forbbiden-key`.

File: testdata/src/key_forbidden/key_forbidden.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import "log/slog"
+
+const (
+	snakeKey = "foo_bar"
+)
+
+func tests() {
+	slog.Info("msg")
+	slog.Info("msg", "foo-bar", 1)
+	slog.Info("msg", "foo_bar", 1)           // want `keys include forbidden values`
```
Comment (mattdowdell) [inline testdata/src/key_forbidden/key_forbidden.go:12]: Can we tell the user which key is forbidden? It could be confusing if I've added 5 attrs.

I also wonder if we should allow the configuration to say why it's forbidden so the user doesn't just slightly change the key name. That would probably require the configuration to be a map which might be more complicated than it's worth.

File: sloglint.go [hunk 2]
```diff
@@ -103,11 +112,13 @@ func flags(opts *Options) flag.FlagSet {
 	boolVar(&opts.NoRawKeys, "no-raw-keys", "enforce using constants instead of raw keys")
 	strVar(&opts.KeyNamingCase, "key-naming-case", "enforce a single key naming convention (snake|kebab|camel|pascal)")
 	boolVar(&opts.ArgsOnSepLines, "args-on-sep-lines", "enforce putting arguments on separate lines")
+	repeatedStrVar(&opts.ForbiddenKeys, "forbidden-key", "enforce not using specific keys. Can be repeated or comma separated")
 
 	return *fset
 }
 
 var slogFuncs = map[string]int{ // funcName:argsPos
+	"log/slog.With":                   0,
```
Comment (tmzane) [inline sloglint.go:121]: Is this change related to the new option?

File: sloglint.go [hunk 3]
```diff
@@ -438,3 +454,26 @@ func argsOnSameLine(fset *token.FileSet, call ast.Expr, keys, attrs []ast.Expr)
...
+	}
+}
+
+func valueChanged(handler func(string) string) func(string) bool {
+	return func(s string) bool {
+		return handler(s) != s
+	}
+}
+
+func trimQuotes(s string) string {
```
Comment (tmzane) [inline sloglint.go:472]: Sorry, why is it needed? I've never seen the use of quoted keys in logger calls.

File: sloglint.go [hunk 4]
```diff
@@ -438,3 +454,26 @@ func argsOnSameLine(fset *token.FileSet, call ast.Expr, keys, attrs []ast.Expr)
...
+
+func forbiddenKeysUsed(forbiddenKeys []string) func(string) bool {
+	return func(s string) bool {
+		// Strip quotes if any.
+		unquoted := trimQuotes(s)
+		return slices.Contains(forbiddenKeys, unquoted)
+	}
+}
+
+func valueChanged(handler func(string) string) func(string) bool {
```
Comment (tmzane) [inline sloglint.go:466]: This looks complicated. Let's leave `badKeyNames` as is and implement the new option as simple `slices.Contains(forbiddenKeys, key)` looping over keys.

File: sloglint_test.go [hunk 5]
```diff
@@ -64,4 +64,9 @@ func TestAnalyzer(t *testing.T) {
 		analyzer := sloglint.New(&sloglint.Options{ArgsOnSepLines: true})
 		analysistest.Run(t, testdata, analyzer, "args_on_sep_lines")
 	})
+
+	t.Run("key foo_bar is forbbiden case", func(t *testing.T) {
```
Comment (tmzane) [inline sloglint_test.go:68]:
```suggestion
	t.Run("forbidden keys", func(t *testing.T) {
```

File: sloglint_test.go [hunk 6]
```diff
@@ -64,4 +64,9 @@ func TestAnalyzer(t *testing.T) {
...
+		analyzer := sloglint.New(&sloglint.Options{ForbiddenKeys: []string{"foo_bar"}})
+		analysistest.Run(t, testdata, analyzer, "key_forbidden")
```
Comment (tmzane) [inline sloglint_test.go:70]:
```suggestion
		analysistest.Run(t, testdata, analyzer, "forbidden_keys")
```

**Review by tmzane — CHANGES_REQUESTED** [review]

File: testdata/src/key_forbidden/key_forbidden.go (hunk 1 above)
Comment (tmzane) [reply testdata/src/key_forbidden/key_forbidden.go:12]: > Can we tell the user which key is forbidden? It could be confusing if I've added 5 attrs.

Agreed, that would be useful. The report message could be something like `foo_bar is forbidden and should not be used`.

> That would probably require the configuration to be a map which might be more complicated than it's worth.

Let's keep it a simple slice for now, we can extend later. If a team needed to document the reason for forbidding certain keys, comments in the config would probably be enough.

File: sloglint.go (hunk 2 above)
Comment (tigrato) [reply sloglint.go:121]: Yes. You can append attributes with the `With` function.

Example: https://go.dev/play/p/PEBgTYRVy2M

```go
package main

import "log/slog"

func main() {
	log := slog.Default()
	log.With("key", "value").Info("test_str", "other_key", "other_value")
}

```

```text
2009/11/10 23:00:00 INFO test_str key=value other_key=other_value
```

Loook that there is `key=value` and `other_key=other_value`

File: sloglint.go (hunk 3 above)
Comment (tigrato) [reply sloglint.go:472]: `getKeyName` returns the `lit.Value` when `lit.Kind=token.String`.

The issue is that `lit.Value` is the literal representation of the string and not the string value. This means, that the `lit.Value` contains the quotes used to define the string. 
This was never an issue because the case converters ignore other chars like `, " and '

File: sloglint.go (hunk 4 above)
Comment (tmzane) [reply sloglint.go:466]: @tigrato UPD. Nevermind, I forgot that we also need to check `Attr` keys. Let's just rename `caseFn` to something more general.

File: sloglint.go (hunk 3 above)
Comment (tmzane) [reply sloglint.go:472]: TIL :) Should we trim quotes before returning from  `getKeyName` then?

File: sloglint.go (hunk 4 above)
Comment (tigrato) [reply sloglint.go]: Done in bce4e31c07313a64d973ac244d41fafb985751d2

File: testdata/src/key_forbidden/key_forbidden.go (hunk 1 above)
Comment (tigrato) [reply testdata/src/key_forbidden/key_forbidden.go]: Updated in bce4e31c07313a64d973ac244d41fafb985751d2.
I quoted the key name to distinguish empty strings

File: sloglint.go (hunk 2 above)
Comment (mattdowdell) [reply sloglint.go:121]: Looks like another escape was recently identified in https://github.com/go-simpler/sloglint/issues/38. Perhaps we can tack the missing `With` onto that and get all the options updated in one pass?

File: sloglint.go [hunk 7]
```diff
@@ -397,6 +414,56 @@ func badKeyNames(info *types.Info, caseFn func(string) string, keys, attrs []ast
...
+				}
+			case 2: // slog.Attr{..., ...} | slog.Attr{Key: ..., Value: ...}
+				expr = attr.Elts[0]
+				if kv1, ok := attr.Elts[0].(*ast.KeyValueExpr); ok && kv1.Key.(*ast.Ident).Name == "Key" {
+					expr = kv1.Value
+				}
+				if kv2, ok := attr.Elts[1].(*ast.KeyValueExpr); ok && kv2.Key.(*ast.Ident).Name == "Key" {
+					expr = kv2.Value
+				}
+			}
```
Comment (mattdowdell) [inline sloglint.go]: Would it be possible (and practical) do this iteratively on the off-chance an extra field gets added to slog.Attr in the future? It would cause all tests for this to fail if that were to happen, but might be a nice edge case. For example:

```suggestion
			for i := range attr.Elts {
				kv, ok := attr.Elts[i].(*ast.KeyValueExpr)
				if !ok {
					continue
				}
				
				if ident, ok := kv.Key.(*ast.Ident); ok && ident.Name == "Key" {
					expr = kv.Value
					break
				}
			}
```

Either way, it might be beneficial to avoid potential panics during the casts.

File: sloglint.go (hunk 2 above)
Comment (tigrato) [reply sloglint.go]: I split it into https://github.com/go-simpler/sloglint/pull/41

File: sloglint.go (hunk 3 above)
Comment (tigrato) [reply sloglint.go:481]: done

File: README.md [hunk 8]
```diff
@@ -167,6 +168,15 @@ slog.Info("a user has logged in",
 )
 ```
 
+### Arguments with forbidden keys
+
+To ensure users don't use reserved words in key-value pairs, you may want to enforce rules to avoid misusages.
+The `forbidden-key` option causes `sloglint` to report usages of forbidden keys. These keys could be typically controlled by slog handlers and its usage should be preserved to avoid key duplication.
```
Comment (mattdowdell) [inline README.md:174]: Can we document how to set this option, e.g. "Keys to forbid can either be comma separated or passed as separate `forbidden-key` options."

**Review by mattdowdell — COMMENTED** [review]
LGTM on the code side, just a few minor comments about the docs.

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (tigrato) [reply README.md:174]: I don't think we should enforce it because by default `slog` doesn't use them - they are used only if you enabled `AddSource` option. 

For everyone that doesn't use them or uses custom handlers, we can break their linter even if they aren't affected by this.

**Comment by tmzane** [comment]
I just remembered that `strconv.Unquote` exists. And it is used in the standard library exactly for this case!

**Review by tmzane — APPROVED** [review]

**Comment by tmzane** [comment]
@tigrato thanks, this is a great feature to have!

@mattdowdell as always, thanks for the review!

**Comment by tmzane** [comment]
Oh, and we should bump Go version in go.mod, because the `slices` package only exists since 1.21.
